### PR TITLE
fix and test for issue #2256

### LINKED ISSUE #2256: error when writing character fields containing NAs to mysql table
I get the following error when trying to write a table to a mysql connection with copy_to():

```
> sanbs_db <- src_mysql("sanbs", host = "localhost", user = "root")
> results_diagnostic.db <- copy_to(sanbs_db, results_diagnostic, temporary = FALSE, indexes = list(c("result_id","donation_id")))
Error in if (n <= 65535) { : missing value where TRUE/FALSE needed
```

I believe the error occurs because I have true NAs in character fields, and the following check in src-mysql.r produces n=NA instead of the expected maximum length:

```
db_data_type.MySQLConnection <- function(con, fields, ...) {
  char_type <- function(x) {
    n <- max(nchar(as.character(x), "bytes"))
    if (n <= 65535) {
      paste0("varchar(", n, ")")
    } else {
      "mediumtext"
    }
}
```

the problem can be demonstrated with:

```
> stringvec <- c("a",NA,"bb")
> max(nchar(stringvec))
[1] NA
> max(na.omit(nchar(stringvec)))
[1] 2
```

I'm not sure what this means for writing NULL values into the db table.

### PR DESCRIPTION
**Pull request by eduardgrebe**
1. encodeString before checking length of character vectors when writing to mysql source

1. rewrite "<NA>" to NA when reading from database tables with SQL

1. test that dataframe containing NAs in character columns writes correctly to database sources

Fixes #2256.

### CODE REVIEW

File: R/src-mysql.r [hunk 1]
```diff
@@ -145,7 +145,9 @@ db_has_table.MySQLConnection <- function(con, table, ...) {
 #' @export
 db_data_type.MySQLConnection <- function(con, fields, ...) {
   char_type <- function(x) {
-    n <- max(nchar(as.character(x), "bytes"))
+    x <- as.character(x)
+    x <- vapply(x, encodeString, character(1))
+    n <- max(nchar(x), "bytes")
```
Comment (krlmlr) [inline R/src-mysql.r]: Wouldn't it be easier to check max(nchar(x[!is.na(x)], "bytes")) ? Please double-check the position of the "bytes" argument.

File: tests/testthat/helper-src.R [hunk 2]
```diff
@@ -5,9 +5,11 @@ if (identical(Sys.info()[["user"]], "hadley")) {
   test_register_src("postgres", src_postgres("test", host = "localhost"))
 } else if (identical(Sys.getenv("TRAVIS"), "true")) {
   test_register_src("postgres", src_postgres("test", user = "travis", password = ""))
+} else if (identical(Sys.info()[["user"]], "eduardgrebe")) {
```
Comment (krlmlr) [inline tests/testthat/helper-src.R]: This doesn't look right to me.

File: tests/testthat/test-tbl-sql.r [hunk 3]
```diff
@@ -9,3 +9,11 @@ test_that("can generate sql tbls with raw sql", {
     expect_equal(collect(tbl), collect(clone))
   }
 })
+
+test_that("NAs in character fields handled by db sources", {
+  df <- data.frame(x=c("a","aa",NA), stringsAsFactors=FALSE)
```
Comment (krlmlr) [inline tests/testthat/test-tbl-sql.r]: Please put a space after commas and around = . A test for NA first and NA in the middle (perhaps in other columns of this data frame) would be great, too.

File: tests/testthat/helper-src.R [hunk 4]
```diff
@@ -5,9 +5,11 @@ if (identical(Sys.info()[["user"]], "hadley")) {
...
+  test_register_src("postgres", src_postgres("test", host = "localhost", user = "eduardgrebe", password = ""))
 }
 
-
+test_register_src("mysql", src_mysql("test", host = "localhost"))
```
Comment (krlmlr) [inline tests/testthat/helper-src.R]: What happens if the MySQL server is not available? It would be great to get a warning (not an error) in this case.

**Comment by krlmlr** [comment]
Thanks. Just a few notes; tests seem to fail on Travis now, too -- we might need to set up MySQL testing on Travis along the lines of what [RMySQL does](https://github.com/rstats-db/RMySQL/blob/master/.travis.yml) (or simply deactivate MySQL on Travis for now).

File: R/src-mysql.r (hunk 1 above)
Comment (eduardgrebe) [inline R/src-mysql.r]: That's what I did initially, but the encodeString applied to the column in the function that writes to the table with SQL results in the string "<NA>" being written - which may be the longest string in the vector - which causes a failure if there are no other strings longer than 4 characters.

File: tests/testthat/helper-src.R (hunk 2 above)
Comment (eduardgrebe) [inline tests/testthat/helper-src.R]: Not sure how to address this, the ifs are false on my machine, meaning no postgres source is initialised.

File: tests/testthat/test-tbl-sql.r (hunk 3 above)
Comment (eduardgrebe) [inline tests/testthat/test-tbl-sql.r]: Good point.

File: R/src-mysql.r (hunk 1 above)
Comment (krlmlr) [inline R/src-mysql.r]: I don't think we should be writing `"<NA>"` for NA values, we should use a SQL `NULL`.

File: tests/testthat/helper-src.R (hunk 2 above)
Comment (krlmlr) [inline tests/testthat/helper-src.R]: Maybe adapt `test_register_src()` to warn instead of throwing an error? Not sure what the implications are.

**Comment by eduardgrebe** [comment]
The main worry I have is that as currently implemented with encodeString in  (I left this unchanged), the string `"<NA>"` is what is stored in the db table, rather than true NULLs. This feels odd to me. Should we be checking for NA values and ensure that NULLs are stored instead?

File: R/src-mysql.r (hunk 1 above)
Comment (eduardgrebe) [inline R/src-mysql.r]: snap. agreed.

**Comment by eduardgrebe** [comment]
I've changed db_insert_into to work like on postgres - NAs now become NULL. Errors are already caught in test_register_src - made the messages clearer, but it shouldn't produce an error when a db engine is not available.

**Comment by krlmlr** [comment]
Thanks. The testthat errors are unrelated (RcppCore/Rcpp#596), but could you please take a look at the other R CMD check messages that Travis shows: https://travis-ci.org/hadley/dplyr/jobs/178453180#L1153 .

File: R/src-mysql.r [hunk 5]
```diff
@@ -145,7 +145,7 @@ db_has_table.MySQLConnection <- function(con, table, ...) {
 #' @export
 db_data_type.MySQLConnection <- function(con, fields, ...) {
   char_type <- function(x) {
-    n <- max(nchar(as.character(x), "bytes"))
+    n <- max(na.omit(nchar(x, "bytes")))
```
Comment (krlmlr) [inline R/src-mysql.r]: max(..., na.rm = TRUE) is slightly better.

File: R/src-mysql.r [hunk 6]
```diff
@@ -181,23 +181,17 @@ db_commit.MySQLConnection <- function(con, ...) {
...
 
-  tmp <- tempfile(fileext = ".csv")
-  utils::write.table(values, tmp, sep = "\t", quote = FALSE, qmethod = "escape",
-    row.names = FALSE, col.names = FALSE)
+  rows <- apply(col_mat, 1, paste0, collapse = ", ")
+  values <- paste0("(", rows, ")", collapse = "\n, ")
 
-  sql <- build_sql("LOAD DATA LOCAL INFILE ", encodeString(tmp), " INTO TABLE ",
-    ident(table), con = con)
+  sql <- build_sql("INSERT INTO ", sql(table), " VALUES ", sql(values))
```
Comment (krlmlr) [inline R/src-mysql.r]: Why are you switching to `INSERT INTO`? Wouldn't that be a lot slower, especially without prepared queries?

I think you could use `\N` as a placeholder for NULL values: http://dev.mysql.com/doc/refman/5.7/en/load-data.html.

File: R/data-temp.r [hunk 7]
```diff
@@ -20,13 +20,15 @@ NULL
 #' @export
 #' @rdname testing
 test_register_src <- function(name, src) {
-  message("Registering testing src: ", name, " ", appendLF = FALSE)
+  message("Registering testing src: ", name)
```
Comment (krlmlr) [inline R/data-temp.r]: Sorry, I forgot how test_register_src() works. The original implementation looks okay to me, I think.

- Doesn't fail
- Prints success and failure messages
- Output: one line only

File: tests/testthat/helper-src.R [hunk 8]
```diff
@@ -1,13 +1,13 @@
 test_register_src("df", src_df(env = new.env(parent = emptyenv())))
 test_register_src("sqlite", src_sqlite(":memory:", create = TRUE))
 
-if (identical(Sys.info()[["user"]], "hadley")) {
-  test_register_src("postgres", src_postgres("test", host = "localhost"))
-} else if (identical(Sys.getenv("TRAVIS"), "true")) {
+if (identical(Sys.getenv("TRAVIS"), "true")) {
```
Comment (krlmlr) [inline tests/testthat/helper-src.R]: Could you please leave this unchanged?

**Comment by krlmlr** [comment]
I've updated master to work around the Rcpp problem, could you please merge/rebase so that testthat tests pass?

**Comment by krlmlr** [comment]
Sorry: I forgot #1975 already resolves this problem. The test, and the enabling of MySQL tests (#870), are still very useful.

**Comment by krlmlr** [comment]
Thanks. I've added your test to #2381.

I see that the current implementation of `db_insert_into()` doesn't work with the development version of RMySQL, due to (at least) rstats-db/RMySQL#117. But this will be fixed eventually, and it currently works with the stable version of RMySQL. I'd rather not switch to using `INSERT INTO`.

**Pull request closed without merging**